nucypher/blockchain/eth/sol/source/contracts/lib/UmbralDeserializer.sol

```diff
@@ -110,20 +129,79 @@ library UmbralDeserializer {
         pointer = copyPoint(pointer, cFrag.pointE1);
         pointer = copyPoint(pointer, cFrag.pointV1);
         cFrag.kFragId = getBytes32(pointer);
-        pointer += BIG_NUM_SIZE;
+        pointer += BIGNUM_SIZE;
         pointer = copyPoint(pointer, cFrag.pointPrecursor);
 
         cFrag.proof = toCorrectnessProof(pointer, cFragBytesLength - CAPSULE_FRAG_SIZE);
     }
 
     /**
-    * @notice Deserialize to pre calculated data
+    * @notice Deserialize to precomputed data
     **/
-    // TODO rename
     function toPreComputedData(bytes memory _preComputedData)
         internal pure returns (PreComputedData memory data)
     {
-        data.data = _preComputedData;
+        require(_preComputedData.length == PRECOMPUTED_DATA_SIZE);
+        uint256 pointer = getPointer(_preComputedData);
+
+        data.pointEyCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointEZxCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointEZyCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointE1yCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointE1HxCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointE1HyCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointE2yCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointVyCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointVZxCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointVZyCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointV1yCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointV1HxCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointV1HyCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointV2yCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointUZxCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointUZyCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointU1yCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointU1HxCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
+
+        data.pointU1HyCoord = uint256(getBytes32(pointer));
+        pointer += BIGNUM_SIZE;
```

Comment:
Is it impossible to clean this up? I would think some sort of looping structure would allow you to iterate over all the `data` values, no?

Reply:
it's ugly because it's assembly code, I think loop will decrease readability because loop will be also assembly with iteration over memory directly

Reply:
Ah, okay. :+1: